python/activity_stream/widget_note.py

```diff
@@ -215,8 +234,98 @@ def apply_thumbnail(self, data):
                 if data["entity"] == reply_widget.created_by:
                     reply_widget.set_thumbnail(image)
 
+    def _add_button_layout(self):
+        """
+        Adds a layout used by the UI buttons.
+        """
+        button_layout = QtGui.QHBoxLayout()
+        self.ui.reply_layout.addLayout(button_layout)
+        button_layout.addStretch(1)
+        button_layout.setSpacing(20)
+        self._button_layout = button_layout
+
+    def add_buttons(self):
+        """
+        Populate the UI with Edit, Cancel, Apply and Reply buttons. Only the Edit
+        and Reply buttons are visible at first.
+        """
+        self._add_button_layout()
+
+        self._edit_button = self.add_edit_button()
+        self._edit_button.clicked.connect(self._on_edit_clicked)
+
+        self._cancel_button = self.add_cancel_button()
+        self._cancel_button.hide()
+        self._cancel_button.clicked.connect(self._on_cancel_clicked)
+
+        self._apply_button = self.add_apply_button()
+        self._apply_button.hide()
+        self._apply_button.clicked.connect(self._on_apply_clicked)
+
+        self._reply_button = self.add_reply_button()
+        self._reply_button.clicked.connect(self._on_reply_clicked)
+
+    def add_edit_button(self):
+        """
+        Adds an initially visible `Edit` button to the UI.
+        """
+        edit_button = ClickableLabel(self)
+        edit_button.setAlignment(QtCore.Qt.AlignLeft|QtCore.Qt.AlignTop)
+        edit_button.setSizePolicy(
+            QtGui.QSizePolicy(
+                QtGui.QSizePolicy.Maximum,
+                QtGui.QSizePolicy.Preferred,
+            )
+        )
+        self._button_layout.addWidget(edit_button)
+        edit_button.setText("Edit")
+        edit_button.setObjectName("edit_button")
+        self._general_widgets.extend([edit_button, self._button_layout])
+        edit_button.clicked.connect(self._on_edit_clicked)
+        return edit_button
+
+    def add_cancel_button(self):
+        """
+        Adds an initially hidden `Edit` button to the UI.
+        """
+        button = ClickableLabel(self)
+        button.setAlignment(QtCore.Qt.AlignLeft|QtCore.Qt.AlignTop)
+        button.setSizePolicy(
+            QtGui.QSizePolicy(
+                QtGui.QSizePolicy.Maximum,
+                QtGui.QSizePolicy.Preferred,
+            )
+        )
+        self._button_layout.addWidget(button)
+        button.setText("Cancel")
+        button.setObjectName("button")
```

Comment:
if you name the button explicitly, i suggest `cancel_button` or something more distinguishable!